File: apps/api/src/http/api_keys.rs
use axum::{
    extract::{Path, State},
    http::HeaderMap,
    Json,
};
use rag_debugger_core::{ApiKey, ApiKeyId, CreateApiKeyRequest, CreatedApiKey};
use uuid::Uuid;

use crate::{auth, error::ApiError, state::AppState};

pub async fn list_api_keys(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> Result<Json<Vec<ApiKey>>, ApiError> {
    let repository = state.repository().ok_or(ApiError::NotReady)?;
    let user =
        auth::authenticate_session(repository.as_ref(), &headers, &state.config().auth).await?;
    Ok(Json(repository.list_api_keys(user.workspace.id).await?))
}

pub async fn create_api_key(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(request): Json<CreateApiKeyRequest>,
) -> Result<Json<CreatedApiKey>, ApiError> {
    let repository = state.repository().ok_or(ApiError::NotReady)?;
    let user =
        auth::authenticate_session(repository.as_ref(), &headers, &state.config().auth).await?;
    let name = request.name.trim();
    if name.is_empty() {
        return Err(ApiError::BadRequest(
            "API key name must not be empty".to_owned(),
        ));
    }
    Ok(Json(
        auth::create_api_key(repository.as_ref(), &user, name.to_owned(), request.scopes).await?,
    ))
}

pub async fn revoke_api_key(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(api_key_id): Path<Uuid>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let repository = state.repository().ok_or(ApiError::NotReady)?;
    auth::authenticate_session(repository.as_ref(), &headers, &state.config().auth).await?;
    repository
        .revoke_api_key(ApiKeyId(api_key_id))
        .await
        .map_err(not_found_to_api("API key"))?;
    Ok(Json(serde_json::json!({ "revoked": true })))
}

fn not_found_to_api(
    label: &'static str,
) -> impl FnOnce(rag_debugger_storage::StorageError) -> ApiError {
    move |error| match error {
        rag_debugger_storage::StorageError::NotFound => {
            ApiError::NotFound(format!("{label} not found"))
        }
        other => ApiError::Storage(other),
    }
}

File: apps/web/src/lib/api/apiKeys.ts
import { jsonRequest, requestJson } from "./client";

export type ApiKeyScope = "ci_eval_runs";

export interface ApiKey {
  id: string;
  workspace_id: string;
  name: string;
  prefix: string;
  scopes: ApiKeyScope[];
  created_at: string;
  last_used_at: string | null;
  revoked_at: string | null;
}

export interface CreateApiKeyRequest {
  name: string;
  scopes?: ApiKeyScope[];
}

export interface CreatedApiKey {
  api_key: ApiKey;
  secret: string;
}

export function listApiKeys(signal?: AbortSignal): Promise<ApiKey[]> {
  return requestJson<ApiKey[]>("/api/v1/api-keys", { signal });
}

export function createApiKey(
  request: CreateApiKeyRequest,
  signal?: AbortSignal,
): Promise<CreatedApiKey> {
  return requestJson<CreatedApiKey>(
    "/api/v1/api-keys",
    jsonRequest("POST", request, signal),
  );
}

export async function revokeApiKey(
  apiKeyId: string,
  signal?: AbortSignal,
): Promise<void> {
  await requestJson<{ revoked: boolean }>(
    `/api/v1/api-keys/${apiKeyId}`,
    jsonRequest("DELETE", {}, signal),
  );
}
